### PR TITLE
PSP-5037 : View / Read Project Details

### PR DESCRIPTION
**Pull request by eddherrera**
image

### CODE REVIEW

File: source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailInformation.tsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  nameSpace?: string;
+}
+
+export const ProjectDetailInformation: React.FunctionComponent<
+  React.PropsWithChildren<IProjectDetailInformationProps>
+> = ({ nameSpace }) => {
+  const { values } = useFormikContext<IProjectForm>();
+  const summary = getIn(values, withNameSpace(nameSpace, 'note'));
+
+  return <Section>This works!{summary}</Section>;
```
Comment (asanchezr) [inline source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailInformation.tsx:16]: Are we sure we want to display "This works!" ?

File: source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailsForm.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+export interface IProjectDetailsFormProps {
+  isEditing?: boolean;
+}
+
+const ProjectDetailsForm: React.FunctionComponent<
+  React.PropsWithChildren<IProjectDetailsFormProps>
+> = () => {
+  const { project } = React.useContext(ProjectStateContext);
+
+  console.log(project);
```
Comment (asanchezr) [inline source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailsForm.tsx:18]: please remove console.log

File: source/frontend/src/features/properties/map/MapRouter.tsx [hunk 3]
```diff
@@ -182,6 +183,16 @@ export const MapRouter: React.FunctionComponent<IMapRouterProps> = memo(props =>
...
+      <AppRoute
+        path={`/mapview/sidebar/project/:id`}
+        customRender={({ match }) => (
+          <ProjectContainer projectId={Number(match.params.id)} onClose={onClose} />
+        )}
+        claim={Claims.PROJECT_VIEW}
+        exact
+        key={'Project'}
+        title={'Project'}
+      />
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/MapRouter.tsx:195]: Please use the dependency injection pattern for Containers -> Views
See an example in **AddFinancialCodeContainer**. We are moving to a new pattern where the container gets the inner view injected via props to help with testing and decoupling.

```tsx
          <AppRoute
            protected
            exact
            path="/admin/financial-code/new"
            customRender={() => (
              <AuthLayout>
                <AddFinancialCodeContainer View={AddFinancialCodeForm} />
              </AuthLayout>
            )}
            role={Roles.SYSTEM_ADMINISTRATOR}
            title={getTitle('Create Financial Code')}
          ></AppRoute>
```

Notice how we "inject" the formik form into the container above. This allows us to create simpler forms for testing so the container tests don't have to rely on knowing about the inner children

File: source/backend/api/Areas/Projects/Controllers/ProjectController.cs [hunk 4]
```diff
@@ -40,6 +40,28 @@ public ProjectController(IProjectService projectService, IMapper mapper)
...
+        [SwaggerOperation(Tags = new[] { "project" })]
+        public async Task<IActionResult> GetById([FromRoute] long id)
+        {
+            var project = await _projectService.GetById(id);
+            if (project is null)
+            {
+                return NotFound();
+            }
+
+            return Ok(_mapper.Map<ProjectModel>(project));
```
Comment (asanchezr) [inline source/backend/api/Areas/Projects/Controllers/ProjectController.cs:62]: Just double-checking that calling the method **Ok()** here is equivalent to returning **JsonResult** ?

Comment (asanchezr) [reply source/backend/api/Areas/Projects/Controllers/ProjectController.cs:62]: According to this [link](https://learn.microsoft.com/en-us/aspnet/core/web-api/advanced/formatting?view=aspnetcore-7.0) there is a subtle difference in returning Ok in that the client can negotiate the return type via the **Accept** header in the request. I'm not opposed to start using Ok but would like that we all agree in a common pattern as I see a mix of Ok and JsonResult even in this same file

File: source/backend/dal/Repositories/ProjectRepository.cs [hunk 5]
```diff
@@ -72,17 +72,17 @@ public Task<Paged<PimsProject>> GetPageAsync(ProjectFilter filter)
...
-            User.ThrowIfNotAuthorized(Permissions.PropertyView);
+            User.ThrowIfNotAuthorized(Permissions.ProjectView);
 
             return await Context.PimsProjects
                     .AsNoTracking()
                     .Include(x => x.ProjectStatusTypeCodeNavigation)
                     .Include(x => x.RegionCodeNavigation)
-                    .Where(x => x.Id == id)
+                    .Where(x => x.Id == projectId)
                     .FirstOrDefaultAsync();
```
Comment (asanchezr) [inline source/backend/dal/Repositories/ProjectRepository.cs:86]: Also not opposed to using **async** DB operations but I think we were not using that previously. Is performance better or is just dev preference to use async EF methods? Would like to sync up with other devs so we so the same

**Review by asanchezr — CHANGES_REQUESTED** [review]
Please update frontend container to follow new practice of dependency injection

File: source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailInformation.tsx (hunk 1 above)
Comment (eddherrera) [reply source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailInformation.tsx]: Removed

File: source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailsForm.tsx (hunk 2 above)
Comment (eddherrera) [reply source/frontend/src/features/projects/detail/projectPages/details/ProjectDetailsForm.tsx]: removed

File: source/frontend/src/features/properties/map/MapRouter.tsx (hunk 3 above)
Comment (eddherrera) [reply source/frontend/src/features/properties/map/MapRouter.tsx]: Updated

File: source/backend/api/Areas/Projects/Controllers/ProjectController.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Areas/Projects/Controllers/ProjectController.cs:62]: I think we should be making an explicit response for the Api, not the content type.

File: source/backend/dal/Repositories/ProjectRepository.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/dal/Repositories/ProjectRepository.cs:86]: thread should perform a little better but overall, not a big win. Async feels a little more modern approach, but we could keep not using it overall.

**Comment by devinleighsmith** [comment]
@asanchezr can you re-review these changes today and clear your flag please?

**Comment by asanchezr** [comment]
> @asanchezr can you re-review these changes today and clear your flag please?

Will do

File: source/backend/dal/Repositories/ProjectRepository.cs (hunk 5 above)
Comment (FuriousLlama) [reply source/backend/dal/Repositories/ProjectRepository.cs:86]: Let's not do this. The request is spawned on its own thread already. So far all calls to this method are just waiting on it, making the async, not a worth having

Comment (FuriousLlama) [reply source/backend/dal/Repositories/ProjectRepository.cs:86]: @eddherrera My PR will change this so no need to worry as we chatted earlier

File: source/backend/tests/unit/api/Services/ProjectServiceTest.cs [hunk 6]
```diff
@@ -196,5 +196,42 @@ public async void Add_Project_Success()
...
+            var service = helper.Create<ProjectService>(user);
+
+            var repository = helper.GetService<Mock<IProjectRepository>>();
+
+            // Act
+            Func<Task> act = async () => await service.GetById(It.IsAny<long>());
+
+            // Assert
+            act.Should().Throw<NotAuthorizedException>();
+            repository.Verify(x => x.Get(It.IsAny<long>()), Times.Never);
```
Comment (asanchezr) [inline source/backend/tests/unit/api/Services/ProjectServiceTest.cs:215]: We are no longer asserting that repos are never called when exception was thrown by service, please remove this line

File: source/frontend/src/features/properties/map/project/add/models.ts [hunk 7]
```diff
@@ -10,13 +10,13 @@ export class ProjectForm {
   region?: NumberFieldValue;
   summary?: string;
 
-  toApi(): Api_Project {
+  toApi(id: number = 1): Api_Project {
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/add/models.ts:13]: sorry, not sure I understand the need to pass the Id here.

File: source/frontend/src/features/properties/map/project/ProjectContainer.test.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import MockAdapter from 'axios-mock-adapter';
+import { mockLookups } from 'mocks';
+import { mockProjectGetResponse } from 'mocks/mockProjects';
+import { lookupCodesSlice } from 'store/slices/lookupCodes';
+import { render, RenderOptions } from 'utils/test-utils';
+
+import { SideBarContextProvider } from '../context/sidebarContext';
+import ProjectContainer, { IProjectContainerViewProps } from './ProjectContainer';
+
+const mockAxios = new MockAdapter(axios);
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/ProjectContainer.test.tsx:11]: Please do not mock axios. In new code we should mock the providers instead (useProjectProvider)

File: source/frontend/src/features/properties/map/project/ProjectContainerView.tsx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import MapSideBarLayout from 'features/mapSideBar/layout/MapSideBarLayout';
+import { useCallback } from 'react';
+import { FaBriefcase } from 'react-icons/fa';
+import styled from 'styled-components';
+
+import ProjectHeader from './common/ProjectHeader';
+import { IProjectContainerViewProps } from './ProjectContainer';
+import ViewSelector from './ViewSelector';
+
+const ProjectContainerView: React.FC<IProjectContainerViewProps> = ({
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/ProjectContainerView.tsx:11]: This view seems to need unit tests

File: source/frontend/src/features/properties/map/project/add/AddProjectContainer.test.tsx [hunk 10]
```diff
@@ -78,42 +84,73 @@ describe('AddProjectContainer component', () => {
...
+    const { getByText, getNameTextbox, getRegionDropdown, getNumberTextbox, getStatusDropdown } =
+      setup();
 
     const formTitle = getByText(/Create Project/i);
-    const input = getNameTextbox();
-    const select = getRegionDropdown();
+    const nameInput = getNameTextbox();
+    const numberInput = getNumberTextbox();
+    const selectRegion = getRegionDropdown();
+    const selectSatus = getStatusDropdown();
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/add/AddProjectContainer.test.tsx:94]: nit: please fix typo selectS(t)atus

File: source/frontend/src/features/properties/map/project/common/ProjectHeader.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+import { Api_Project } from 'models/api/Project';
+import { Col, Row } from 'react-bootstrap';
+import styled from 'styled-components';
+import { prettyFormatDate } from 'utils';
+
+export interface IProjectHeaderProps {
+  project?: Api_Project;
+}
+
+const ProjectHeader: React.FunctionComponent<React.PropsWithChildren<IProjectHeaderProps>> = ({
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/common/ProjectHeader.tsx:12]: Header component should have tests

File: source/frontend/src/features/properties/map/project/update/UpdateProjectContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+export interface IUpdateProjectContainerProps {
+  initialValues: ProjectForm;
+  projectStatusOptions: SelectOption[];
+  projectRegionOptions: SelectOption[];
+  validationSchema?: any | (() => any);
+  onSubmit: (values: ProjectForm, formikHelpers: FormikHelpers<ProjectForm>) => void | Promise<any>;
+}
+
+const UpdateProjectContainer = React.forwardRef<FormikProps<any>, IUpdateProjectContainerProps>(
```
Comment (asanchezr) [inline source/frontend/src/features/properties/map/project/update/UpdateProjectContainer.tsx:19]: This doesn't seem relevant to this PR but I guess is fine to keep it since you will working with this in your next story/PR

File: source/frontend/src/mocks/mockProjects.ts [hunk 13]
```diff
@@ -33,22 +33,52 @@ export const mockProjects = (): Api_Project[] => [
 
 export const mockProjectPostResponse = (
   id: number = 1,
```
Comment (asanchezr) [inline source/frontend/src/mocks/mockProjects.ts:35]: Not sure we want to send Id = 1 to create a new project. I believe the existing pattern is to leave Id undefined for POSTs

File: source/frontend/src/models/api/Project.ts [hunk 14]
```diff
@@ -1,13 +1,14 @@
...
 
 export interface Api_Project extends Api_ConcurrentVersion, Api_AuditFields {
   id?: number;
   projectStatusTypeCode?: Api_TypeCode<string>;
-  businessFunctionCode?: any; // TODO: Match new code types
-  costTypeCode?: any; // TODO: Match new code types
-  workActivityCode?: any; // TODO: Match new code types
+  businessFunctionCode?: Api_CodeType;
+  costTypeCode?: Api_CodeType;
+  workActivityCode?: Api_CodeType;
```
Comment (asanchezr) [inline source/frontend/src/models/api/Project.ts:11]: They look almost identical but I believe Api_CodeType is specific to Country and Provinces. It looks that here the more relevant model would be Api_FinancialCode which was created for business function, cost type, etc

**Review by asanchezr — APPROVED** [review]

File: source/frontend/src/features/properties/map/project/common/ProjectHeader.tsx (hunk 11 above)
Comment (asanchezr) [reply source/frontend/src/features/properties/map/project/common/ProjectHeader.tsx:12]: a single snapshot test would be sufficient
